### PR TITLE
s3-style URL fix

### LINKED ISSUE #31093: s3-style URL does not work.
**Describe the issue**

This works successfully:
```
aws s3 cp 'hits.csv' 's3://milovidov-clickhouse-test/hits.csv'
```

This does not:
```
SELECT count() FROM s3('s3://milovidov-clickhouse-test/hits.csv', '...', '...', 'CSV', 'WatchID UInt64, JavaEnable UInt8, Title String, GoodEvent Int16, EventTime DateTime, EventDate Date, CounterID UInt32, ClientIP UInt32, RegionID UInt32, UserID UInt64, CounterClass Int8, OS UInt8, UserAgent UInt8, URL String, Referer String, Refresh UInt8, RefererCategoryID UInt16, RefererRegionID UInt32, URLCategoryID UInt16, URLRegionID UInt32, ResolutionWidth UInt16, ResolutionHeight UInt16, ResolutionDepth UInt8, FlashMajor UInt8, FlashMinor UInt8, FlashMinor2 String, NetMajor UInt8, NetMinor UInt8, UserAgentMajor UInt16, UserAgentMinor FixedString(2), CookieEnable UInt8, JavascriptEnable UInt8, IsMobile UInt8, MobilePhone UInt8, MobilePhoneModel String, Params String, IPNetworkID UInt32, TraficSourceID Int8, SearchEngineID UInt16, SearchPhrase String, AdvEngineID UInt8, IsArtifical UInt8, WindowClientWidth UInt16, WindowClientHeight UInt16, ClientTimeZone Int16, ClientEventTime DateTime, SilverlightVersion1 UInt8, SilverlightVersion2 UInt8, SilverlightVersion3 UInt32, SilverlightVersion4 UInt16, PageCharset String, CodeVersion UInt32, IsLink UInt8, IsDownload UInt8, IsNotBounce UInt8, FUniqID UInt64, OriginalURL String, HID UInt32, IsOldCounter UInt8, IsEvent UInt8, IsParameter UInt8, DontCountHits UInt8, WithHash UInt8, HitColor FixedString(1), LocalEventTime DateTime, Age UInt8, Sex UInt8, Income UInt8, Interests UInt16, Robotness UInt8, RemoteIP UInt32, WindowName Int32, OpenerName Int32, HistoryLength Int16, BrowserLanguage FixedString(2), BrowserCountry FixedString(2), SocialNetwork String, SocialAction String, HTTPError UInt16, SendTiming UInt32, DNSTiming UInt32, ConnectTiming UInt32, ResponseStartTiming UInt32, ResponseEndTiming UInt32, FetchTiming UInt32, SocialSourceNetworkID UInt8, SocialSourcePage String, ParamPrice Int64, ParamOrderID String, ParamCurrency FixedString(3), ParamCurrencyID UInt16, OpenstatServiceName String, OpenstatCampaignID String, OpenstatAdID String, OpenstatSourceID String, UTMSource String, UTMMedium String, UTMCampaign String, UTMContent String, UTMTerm String, FromTag String, HasGCLID UInt8, RefererHash UInt64, URLHash UInt64, CLID UInt32')
```

It argues about the URL:

```
Code: 36. DB::Exception: Bucket or key name are invalid in S3 URI. (BAD_ARGUMENTS)
```

But I'm using the same URL as with `s3 aws` tool :(

### PR DESCRIPTION
**Pull request by yariks5s**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Allow S3-style URLs for table functions `s3`, `gcs`, `oss`.  URL is automatically converted to HTTP. Example:  `'s3://clickhouse-public-datasets/hits.csv'` is converted to `'https://clickhouse-public-datasets.s3.amazonaws.com/hits.csv'`

Closes #31093.
### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
*This is an automated comment for commit e1f40adfe272b5f67279dc6cc2f197d28b24ae97 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/54931/e1f40adfe272b5f67279dc6cc2f197d28b24ae97/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>✅ success</td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>⏳ pending</td></tr>
<tbody>
</table>

**Comment by alexey-milovidov** [comment]
@yariks5s, this is a new feature, not "not for changelog."

File: src/IO/S3/URI.cpp [hunk 1]
```diff
@@ -18,6 +18,26 @@
 namespace DB
 {
 
+struct UriConverter
```
Comment (alexey-milovidov) [inline src/IO/S3/URI.cpp:21]: URI

File: src/IO/S3/URI.cpp [hunk 2]
```diff
@@ -18,6 +18,26 @@
...
+            domain = S3_DOMAIN;
+        else if (uri.getScheme() == "oss")
+            domain = OSS_DOMAIN;
+        else if (uri.getScheme() == "gs")
+            domain = GCS_DOMAIN;
+        uri.setScheme("https");
+        uri.setHost(uri.getHost() + domain);
+    }
+private:
+    static constexpr auto S3_DOMAIN = ".s3.amazonaws.com";
```
Comment (alexey-milovidov) [inline src/IO/S3/URI.cpp:36]: When a constant is used once, it's ok to have it inline, without a separate constant definition.

**Comment by alexey-milovidov** [comment]
Need to add a test.

**Comment by alexey-milovidov** [comment]
By the way, we can merge the first version without config and add config in a subsequent PR.

**Comment by alexey-milovidov** [comment]
@yariks5s let's check - is it true that AWS SDK doesn't support s3-style URLs?

**Comment by yariks5s** [comment]
> @yariks5s let's check - is it true that AWS SDK doesn't support s3-style URLs?

I just checked docs [here](https://docs.aws.amazon.com/pdfs/sdk-for-cpp/v1/developer-guide/aws-sdk-cpp-dg.pdf) and could not find anything regarding s3-style links, but at page 108, in the bottom, I found a host with an appropriate ending, so it is probably not supported but I cannot be 100% sure.

**Comment by alexey-milovidov** [comment]
Please check the test results.

**Review by KochetovNicolai — APPROVED** [review]

**Review by alexey-milovidov — APPROVED** [review]

**Pull request merged**